### PR TITLE
[LTC] Implement Eowyn, Shieldmaiden

### PR DESCRIPTION
**Pull request by Susucre**
The watcher is a little tricky, as we want to watch for all humans entering for each player. Then, the condition check will actually check if another human actually entered under the controller's control.

### CODE REVIEW

**Comment by Susucre** [comment]
[[Eowyn, Shieldmaiden]]

File: Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        // Then if you control six or more Humans, draw a card.
+
+        TriggeredAbility triggeredAbility = new BeginningOfCombatTriggeredAbility(
+            Zone.BATTLEFIELD,
+            new CreateTokenEffect(new HumanKnightToken(), 2),
+            TargetController.YOU, false, false
+        );
+        triggeredAbility.addEffect(new ConditionalOneShotEffect(
+            new DrawCardSourceControllerEffect(1),
+            new PermanentsOnTheBattlefieldCondition(filter, ComparisonType.MORE_THAN, 5)
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java:58]: Need to make sure this checks controlled Humans, not any Humans

Comment (Susucre) [reply Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java:58]: I can add a true boolean as 4th argument, but the PermanentsOnTheBattlefieldCondition constructor with 3 arguments is already checking only controlled permanents.

https://github.com/magefree/mage/blob/7a33ca812ce37ecd2651b34aee95e3b95b4be05c/Mage/src/main/java/mage/abilities/condition/common/PermanentsOnTheBattlefieldCondition.java#L49C8-L59

Comment (xenohedron) [reply Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java:58]: Ah never mind then, sorry

File: Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    boolean hasPlayerHadAnotherHumanEnterThisTurn(Permanent sourcePermanent, UUID playerId) {
+        // The trick there is that we need to exclude the sourcePermanent of the trigger (Eowyn most likely),
+        //     from the humans that entered the battlefield this turn under the player.
+        UUID sourceId = null;
+        if(sourcePermanent != null)
+            sourceId = sourcePermanent.getId();
+
+        UUID finalSourceId = sourceId;
```
[the remark this answers is not in the document]
Comment (Susucre) [reply Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java:135]: I can not use a non-final value inside the `anyMatch` lambda expression.
I could go for an inline ternary instead:
> UUID sourceId = sourcePermanent == null ? null : sourcePermanent.getId();

Would that be better?

File: Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            && watcher.hasPlayerHadAnotherHumanEnterThisTurn(
+                source.getSourcePermanentOrLKI(game),
+                source.getControllerId());
+    }
+}
+
+class EowynShieldmaidenWatcher extends Watcher {
+
+    // Map players UUID to theirs humans UUID that entered this turn.
+    private final Map<UUID, Set<UUID>> humanEnterings = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (Susucre) [reply Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java:98]: Uhm indeed.
What is the easier way to identify the Eowyn Permanent and lose identification on next flicker (or reanimation, or anything really)? UUID + zcc? We certaintly have something for that kind of identification already set up?

[the remark this answers is not in the document]
Comment (xenohedron) [reply Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java:98]: Well I can't really call it easy, but we do have a thing called MageObjectReference for such purposes

File: Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (xenohedron) [reply Mage.Sets/src/mage/cards/e/EowynShieldmaiden.java:135]: Ah that explains it. I'm personally a fan of the ternary operator.

**Comment by Susucre** [comment]
Alright, this is properly working using MOR and not UUID to identify the entering humans.

There was something quite weird with simulations triggering the watcher, but with distinct MOR identifiers.
I did add a !game.isSumulation() to only register the proper entering events.

**Review by xenohedron — APPROVED** [review]

**Pull request merged**